### PR TITLE
feat(benchmarks) Add LLM evaluation pipeline for Finance challenge

### PR DESCRIPTION
**Pull request by yan-gao-GY**
## Proposal

Add evaluation code of Finance challenge for FlowerTune LLM leaderboard.

### CODE REVIEW

File: benchmarks/flowertune-llm/evaluation/finance/README.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+## Evaluation for Finance challenge
```
Comment (jafermarq) [inline benchmarks/flowertune-llm/evaluation/finance/README.md:1]:
```suggestion
# Evaluation for Finance challenge
```

File: benchmarks/flowertune-llm/evaluation/finance/README.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+We leverage the sentiment classification pipeline on finance-related text provided by [FinGPT](https://github.com/AI4Finance-Foundation/FinGPT/tree/master) to evaluate our fine-tuned LLMs.
+Three datasets have been selected for this evaluation: [FPB](https://huggingface.co/datasets/takala/financial_phrasebank), [FIQA](https://huggingface.co/datasets/pauri32/fiqa-2018), and [TFNS](https://huggingface.co/datasets/zeroshot/twitter-financial-news-sentiment). 
+
+
+### Step 0. Set up Environment
```
Comment (jafermarq) [inline benchmarks/flowertune-llm/evaluation/finance/README.md:7]:
```suggestion
## Environment Setup
```

File: benchmarks/flowertune-llm/evaluation/finance/README.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+```shell
+git clone --depth=1 https://github.com/adap/flower.git && mv flower/benchmarks/flowertune-llm/evaluation/finance ./flowertune-eval-finance && rm -rf flower && cd flowertune-eval-finance
+```
+
+Then, install dependencies with:
```
Comment (jafermarq) [inline benchmarks/flowertune-llm/evaluation/finance/README.md:13]:
```suggestion
Create a new Python environment (we recommend Python 3.10), activate it, then install dependencies with:
```

File: benchmarks/flowertune-llm/evaluation/finance/README.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+```shell
+# From a new python environment, run:
+pip install -e .
+
+# Log in HuggingFace account
+huggingface-cli login
+```
+
+### Step 1. Generate model decision & calculate accuracy
```
Comment (jafermarq) [inline benchmarks/flowertune-llm/evaluation/finance/README.md:23]:
```suggestion
## Generate model decision & calculate accuracy
```

File: benchmarks/flowertune-llm/evaluation/finance/README.md [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+--peft-path=/path/to/fine-tuned-peft-model-dir/ # e.g., ./peft_1
+--run-name=fl  # arbitrary name for this run  
+--batch-size=32 
+--quantization=4 
+--datasets=fpb,fiqa,tfns
+```
+The model answers and accuracy values will be saved to `benchmarks/generation_{dataset_name}_{run_name}.jsonl` and `benchmarks/acc_{dataset_name}_{run_name}.txt`, respectively.
+
+> [!NOTE]
+> Please ensure that you provide all **three accuracy values** for three evaluation datasets when submitting to the LLM Leaderboard (see the [`Make Submission`](https://github.com/adap/flower/tree/main/benchmarks/flowertune-llm/evaluation#make-submission-on-flowertune-llm-leaderboard) section).
```
Comment (jafermarq) [inline benchmarks/flowertune-llm/evaluation/finance/README.md:36]:
```suggestion
> Please ensure that you provide all **three accuracy values (FPB, FIQA, TFNS)** for three evaluation datasets when submitting to the LLM Leaderboard (see the [`Make Submission`](https://github.com/adap/flower/tree/main/benchmarks/flowertune-llm/evaluation#make-submission-on-flowertune-llm-leaderboard) section).
```

File: benchmarks/flowertune-llm/evaluation/finance/eval.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
+# This python file is adapted from https://github.com/AI4Finance-Foundation/FinGPT/blob/master/fingpt/FinGPT_Benchmark/benchmarks/benchmarks.py
+
+import argparse
+
+import torch
+from fiqa import test_fiqa
+from fpb import test_fpb
+from nwgi import test_nwgi
```
Comment (jafermarq) [inline benchmarks/flowertune-llm/evaluation/finance/eval.py:8]: this file is not in the directory.

File: benchmarks/flowertune-llm/evaluation/finance/eval.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+with torch.no_grad():
+    for dataset in args.datasets.split(","):
+        if dataset == "fpb":
+            test_fpb(args, model, tokenizer)
+        elif dataset == "fiqa":
+            test_fiqa(args, model, tokenizer)
+        elif dataset == "tfns":
+            test_tfns(args, model, tokenizer)
+        elif dataset == "nwgi":
+            test_nwgi(args, model, tokenizer)
```
Comment (jafermarq) [inline benchmarks/flowertune-llm/evaluation/finance/eval.py:67]: remove?

File: benchmarks/flowertune-llm/evaluation/finance/pyproject.toml [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+authors = [
+    { name = "The Flower Authors", email = "hello@flower.ai" },
+]
+dependencies = [
+    "peft==0.6.2",
+    "scikit-learn==1.5.0",
+    "datasets==2.20.0",
+    "sentencepiece==0.2.0",
+    "protobuf==5.27.1",
+    "bitsandbytes==0.43.1",
```
Comment (jafermarq) [inline benchmarks/flowertune-llm/evaluation/finance/pyproject.toml:18]: are these dependencies the same as in the `flwr new` template?

**Review by jafermarq — COMMENTED** [review]
@yan-gao-GY , some suggestions in the readme just to align it to the recent changes to the general-nlp evaluation

File: benchmarks/flowertune-llm/evaluation/finance/benchmarks.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+    # Run inference
+    dataset, acc = inference(dataset, model, tokenizer, batch_size)
+
+    # Save results and generations
+    save_results(name, run_name, dataset, acc)
+
+
+def infer_fpb(model, tokenizer, batch_size, run_name):
+    name = "fpb"
+    dataset = load_data("financial_phrasebank", "sentences_50agree")
```
Comment (jafermarq) [inline benchmarks/flowertune-llm/evaluation/finance/benchmarks.py:40]:
```suggestion
    dataset = load_data("takala/financial_phrasebank", "sentences_50agree")
```

**Review by jafermarq — APPROVED** [review]
